merkle-tree/bounded-vec/src/lib.rs

```diff
@@ -520,31 +522,293 @@ where
+    /// * `^` - first element
+    /// * `$` - last element
+    /// * `#` - visited elements
+    ///
+    /// Length: 6
+    /// Capacity: 8
+    /// First index: 0
+    /// Last index: 5
+    ///
+    /// Start iteration from: 2
+    ///
+    /// Should iterate over elements from 2 to 5, with 4 iterations.
+    #[test]
+    fn test_cyclic_bounded_vec_iter_from_without_reset_not_full() {
+        let mut cyclic_bounded_vec = CyclicBoundedVec::with_capacity(8);
+
+        for i in 0..6 {
+            cyclic_bounded_vec.push(i);
+        }
+
+        assert_eq!(cyclic_bounded_vec.len(), 6);
+        assert_eq!(cyclic_bounded_vec.capacity(), 8);
+        assert_eq!(cyclic_bounded_vec.first_index, 0);
+        assert_eq!(cyclic_bounded_vec.last_index, 5);
+
+        let elements = cyclic_bounded_vec.iter_from(2).collect::<Vec<_>>();
+        assert_eq!(elements.len(), 4);
+        assert_eq!(elements.as_slice(), &[&2, &3, &4, &5]);
+    }
+
+    /// Iteration without reset in a vector which is not full.
+    ///
+    /// ```
+    ///              #  #  #  #  #  #
+    ///        ^                    $
+    /// index [0, 1, 2, 3, 4, 5, 6, 7]
+    /// value [0, 1, 2, 3, 4, 5, 6, 7]
+    /// ```
+    ///
+    /// * `^` - first element
+    /// * `$` - last element
+    /// * `#` - visited elements
+    ///
+    /// Length: 8
+    /// Capacity: 8
+    /// First index: 0
+    /// Last index: 7
+    ///
+    /// Start iteration from: 2
+    ///
+    /// Should iterate over elements 2..7 - 6 iterations.
+    #[test]
+    fn test_cyclic_bounded_vec_iter_from_without_reset_full() {
+        let mut cyclic_bounded_vec = CyclicBoundedVec::with_capacity(8);
+
+        for i in 0..8 {
+            cyclic_bounded_vec.push(i);
+        }
+
+        assert_eq!(cyclic_bounded_vec.len(), 8);
+        assert_eq!(cyclic_bounded_vec.capacity(), 8);
+        assert_eq!(cyclic_bounded_vec.first_index, 0);
+        assert_eq!(cyclic_bounded_vec.last_index, 7);
+
+        let elements = cyclic_bounded_vec.iter_from(2).collect::<Vec<_>>();
+        assert_eq!(elements.len(), 6);
+        assert_eq!(elements.as_slice(), &[&2, &3, &4, &5, &6, &7]);
+    }
+
+    /// Iteration with reset.
+    ///
+    /// Insert elements over capacity, so the vector resets and starts
+    /// overwriting elements from the start - 12 elements into a vector with
+    /// capacity 8.
+    ///
+    /// The resulting data structure looks like:
+    ///
+    /// ```
+    ///        #  #   #   #        #  #
+    ///                   $  ^
+    /// index [0, 1,  2,  3, 4, 5, 6, 7]
+    /// value [8, 9, 10, 11, 4, 5, 6, 7]
+    /// ```
+    ///
+    /// * `^` - first element
+    /// * `$` - last element
+    /// * `#` - visited elements
+    ///
+    /// Length: 8
+    /// Capacity: 8
+    /// First: 4
+    /// Last: 3
+    ///
+    /// Start iteration from: 6
+    ///
+    /// Should iterate over elements 6..7 and 8..11 - 6 iterations.
+    #[test]
+    fn test_cyclic_bounded_vec_iter_reset() {
```

Comment:
I am missing a test for the edge case that we need to cover in the Merkle tree.
Consider an example:
// capacity 5 [0,1,2,3,4]
// start with index 2 and want to do 4 iterations
```
let elements = cyclic_bounded_vec.iter_from(2).collect::<Vec<_>>();
assert_eq!(elements.as_slice(), &[&3, &4, &0, &1]);
 ```

Reply:
Sure, I can add it

Reply:
Your expected_result is not correct for such example though. The `[0, 1, 2, 3, 4]` vector didn't cycle over yet. Starting from 2 should result in `[2, 3, 4]`. 0 and 1 are elements which are older than than 2. Why would you want to iterate over changelog entries which predate your requested index? We want to use only newer changelogs, not older ones.

That's exactly why I'm keeping track of `first_index` and `last_index` - to figure out if I really need to jump to the beginning. These counters are making the algorithm aware of whether the vector already cycled or not.

Your example would be somehow correct if we did an explicit push of `0` and `1´ elements afterwards and the vector would cycle over. I'm always incrementing pushed elements by 1 though, just to make it crystal clear which elements are newer and older.

Reply:
I added a test for the input you mentioned, but with output which IMO is correct. LMK WDYT :)

Reply:
fair I didn't intend to put any meaning into the numbers itself.
And I got the two wrong it should have been included since we are starting from index two inclusive.
``` 
// capacity 5 [43,22,55,13,42]
// using first and last index these would be: first 2, last 1
// without first and last, self.current_index would be 1, and the user requests to start at index 2
let elements = cyclic_bounded_vec.iter_from(2);
let expected_result = [55,13,42, 43, 22];
for i in 0..4 {
    assert_eq!(elements.next(), expected_result[i]);
}
```